crypto/engine/eng_list.c

```diff
@@ -138,17 +139,16 @@ ENGINE *ENGINE_get_first(void)
 
     CRYPTO_THREAD_write_lock(global_engine_lock);
     ret = engine_list_head;
-    if (ret) {
-        ret->struct_ref++;
-        engine_ref_debug(ret, 0, 1);
-    }
     CRYPTO_THREAD_unlock(global_engine_lock);
+    if (ret != NULL)
+        ENGINE_UP_STRUCT_REF(ret, &ref_cnt);
```

Comment:
So what if another thread frees the last reference to this engine between lines 142 and 143 of this thread?

Reply:
This code and others like it isn't making much sense to me. Why is ref_cnt a local (stack based) variable? Once you leave this function you lost your counter. It's also not initialized.

Reply:
I was of course total confused, ref_cnt is not what's been modified, it's ret->strut_ref that gets modified and ref_cnt is the new value.